File: src/main/java/aoc2019/day13/Day13.java
package aoc2019.day13;


import aoc2019.day09.Output;

import java.awt.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import static aoc2019.day09.Day9Kt.processParameterMode;


public class Day13 {

    static Map<Point, Tile> processInput(ArrayList<Long> numbers) {

        Map<Point, Tile> tileMap = new HashMap<>();
        int i = 0;
        int relativeBase = 0;
        StringBuilder outputBuilder = new StringBuilder();

        while (i < numbers.size()) {
            int opCode = Math.toIntExact(numbers.get(i));
            if (opCode == 99) {
                break;
            }

            Output output = processParameterMode(numbers, i, opCode, 0, relativeBase);
            if (!output.getCode().equals("")) {
                outputBuilder.append(output.getCode()).append(",");
            }

            if (output.getRelativeBase() != 0) {
                relativeBase = output.getRelativeBase();
            }
            i += output.getIndex();
        }

        String[] outputs = outputBuilder.toString().split(",");

        for (i = 0; i < outputs.length; i += 3) {
            int x = Integer.parseInt(outputs[i]);
            int y = Integer.parseInt(outputs[i + 1]);
            int tileId = Integer.parseInt(outputs[i + 2]);
            tileMap.put(new Point(x, y), Tile.valueOf(tileId));
        }

        return tileMap;
    }

    static void printTileMap(Map<Point, Tile> tileMap) {
        int minX = tileMap.keySet().stream().mapToInt(p -> p.x).min().orElse(-1);
        int maxX = tileMap.keySet().stream().mapToInt(p -> p.x).max().orElse(-1);
        int minY = tileMap.keySet().stream().mapToInt(p -> p.y).min().orElse(-1);
        int maxY = tileMap.keySet().stream().mapToInt(p -> p.y).max().orElse(-1);

        for (int y = minY; y <= maxY; y++) {
            for (int x = minX; x <= maxX; x++) {
                Point key = new Point(x, y);
                if (tileMap.containsKey(key)) {
                    System.out.print(tileMap.get(key).getSymbol());
                } else {
                    System.out.print(" ");
                }
            }
            System.out.println();
        }
    }

    static int countTilesOfType(Tile tileType, Map<Point, Tile> processInput) {
        return (int) processInput.values().stream().filter(tile -> tile.equals(tileType)).count();
    }
}

File: 2015-2023/src/main/kotlin/aoc2019/day13/Day13.kt
package aoc2019.day13

import aoc2019.day09.processParameterMode
import aoc2019.day13.Tile.Companion.valueOf
import java.awt.Point
import java.util.*

fun processInput(numbers: ArrayList<Long>): Map<Point, Tile?> {
    val tileMap: MutableMap<Point, Tile?> = HashMap()
    var i = 0
    var relativeBase = 0
    val outputBuilder = StringBuilder()
    while (i < numbers.size) {
        val opCode = Math.toIntExact(numbers[i])
        if (opCode == 99) {
            break
        }
        val output = processParameterMode(numbers, i, opCode, 0, relativeBase)
        if (output.code != "") {
            outputBuilder.append(output.code).append(",")
        }
        if (output.relativeBase != 0) {
            relativeBase = output.relativeBase
        }
        i += output.index
    }
    val outputs = outputBuilder.toString().split(",").toTypedArray()
    i = 0
    while (i < outputs.size-1) {
        val x = outputs[i].toInt()
        val y = outputs[i + 1].toInt()
        val tileId = outputs[i + 2].toInt()
        if (valueOf(tileId) != null)
            tileMap[Point(x, y)] = valueOf(tileId)
        i += 3
    }
    return tileMap
}

fun printTileMap(tileMap: Map<Point, Tile?>) {
    val minX = tileMap.keys.stream().mapToInt { p: Point -> p.x }.min().orElse(-1)
    val maxX = tileMap.keys.stream().mapToInt { p: Point -> p.x }.max().orElse(-1)
    val minY = tileMap.keys.stream().mapToInt { p: Point -> p.y }.min().orElse(-1)
    val maxY = tileMap.keys.stream().mapToInt { p: Point -> p.y }.max().orElse(-1)
    for (y in minY..maxY) {
        for (x in minX..maxX) {
            val key = Point(x, y)
            if (tileMap.containsKey(key)) {
                print(tileMap[key]!!.symbol)
            } else {
                print(" ")
            }
        }
        println()
    }
}

fun countTilesOfType(tileType: Tile, processInput: Map<Point, Tile>): Int {
    return processInput.values.stream().filter { tile: Tile -> tile == tileType }.count()
        .toInt()
}
